### DISCUSSION TITLE
Where are the default error page templates?

### DISCUSSION THREAD

**Discussion by chen-gloria** [symfony/symfony#44425] [Q&A] [answered]
In this documentation: https://symfony.com/doc/current/controller/error_pages.html I was looking for the templates of the default error page. I installed the TwigBundle `composer require symfony/twig-pack`, but I couldn't find the `error.html.twig` or any other exception/error twigs according to this documentation.

I tried to create `templates/bundles/TwigBundle/Exception/` directory to overwrite the `error.html.twig` or `error404.html.twig` or `exception.html.twig`. None of them work: None of my customised `error.html.twig` or `error404.html.twig` or `exception.html.twig` page displayed.

Is there an issue with the documentation or should I look for some other directories?

**Comment by noniagriconomie** [comment]
the default source files are here https://github.com/symfony/error-handler/tree/5.4/Resources/views
it should work with twig override, `None of them work.` can you be more precise?

**Reply by chen-gloria** [reply]
@noniagriconomie I turned off the DEBUG mode, it works fine now! It seems like it was just I didn't turn off the dev environment. I added `APP_DEBUG=0` and it works now. Thanks!

**Comment by BahmanMehrdad** [comment, marked as answer]
See this link:

https://symfony.com/doc/current/controller/error_pages.html#testing-error-pages-during-development

**Reply by chen-gloria** [reply]
Thanks! I tried again. It does work. I use Symfony 5.3 and my `framework.yaml` is:

```
when@dev:
    _errors:
        resource: '@FrameworkBundle/Resources/config/routing/errors.xml'
        prefix: /_error
```

And when I tried to test my error page in dev. I just go to `http://localhost/xxxx(my project URL name)/_error/{statusCode}`.

Note that sometimes you need to add `your project name` or anything you may need to add in your project URL after the `localhost`, and then add `/_error/{statusCode}` to get the error HTML page. In my case, I use SimpleSAML. I have different names for different apps to login, so I need to add `/xxxx(my project name)` after `/localhost` and then `/_error/{statusCode}` to get the error page.

By going to `http://localhost/xxxx(my project URL name)/_error/{statusCode}`, you don't need to set `APP_DEBUG=0` eveytime.

Since I use Docker in my project. A downside that I've found for setting `APP_DEBUG=0` in Docker environment is that you have to do `docker compose down` and then `docker compose up` every time after you change your `APP_DEBUG` to `0` (non-dev) or change it back to dev --- I found that I'm not able to clear cache when `APP_DEBUG=0`, so I can't see the changes in my error page unless I do `docker compose down` and `docker compose up` again, which is very annoying.

I've tried different ways to clear the cache according to this [doc](https://symfony.com/doc/5.3/cache.html#clearing-the-cache), none of them actually clear the cache for me. Actually, `php bin/console cache:clear` or other cache clear commands doesn't really pick up the changes of the `APP_DEBUG` when I tried to change `APP_DEBUG` to `0`. It still told me that `debug=true` even though I've changed `APP_DEBUG` to `0`. And only if I do `docker compose down` and then `docker compose up` again, it will update my `APP_DEBUG`.

Even though I haven't tested the customized page in my prod environment yet, I can access my customised page by just going to `http://localhost/xxxx(my project URL name)/_error/{statusCode}` -- It works for me for now.

I will update if it works in my `prod` next time.

Thanks, @BahmanMD  @noniagriconomie and Symfony community for helping me out in this.

**Reply by chen-gloria** [reply]
**Update 1:** I found that the reason why I had to do `docker compose down` and `docker compose up` again to update my `APP_DEBUG` status is because I have a `.env.local.php` file which contains the `APP_DEBUG => '0'` which overwrites `.env` since it has higher hierarchy according to this [doc](https://symfony.com/doc/current/configuration.html#overriding-environment-values-via-env-local). And stupidly I kept updating my `APP_DEBUG` in `.env` rather than `.env.local.php` where I should have updated into.

I actually prefer the way of changing the `APP_DEBUG` to test my error page since it actually gives me the correct simulation of the user's behaviour, which is very important for my app development. For example, I can just enter a URL, pretending to be a user, and see what error page will return, or what the user should get, or what the user would expect to get after the error page, rather than going to `http://localhost/xxxx(my project URL name)/_error/{statusCode}` page.

So I guess there are 2 ways to test your error page eventually:

### 1- Go to `http://localhost/xxxx([OPTIONAL]possible project URL name)/_error/{statusCode}` to test your error page

Do as what the [doc](https://symfony.com/doc/current/controller/error_pages.html#testing-error-pages-during-development) said, setup your `framework.yaml` file, overwrite the Twig and test in `http://localhost/xxxx(my project URL name)/_error/{statusCode}`

It gives you a convenient way to access the Twig template or your error pages that you just modified. However, it won't give you a similar simulation of the user's behavior.

### 2- Change `APP_DEBUG` and test in non-dev environment

Change your `APP_DEBUG` in whatever `.env` or `.env.local.php` or any environment file, in which you can still clear your cache any time, just pay attention to the file hierarchy according to this [doc](https://symfony.com/doc/current/configuration.html#overriding-environment-values-via-env-local).

In this way, you can easily test your error page and also simulate your user's behavior in terms of what they could see in the URL they would try to enter, maybe to improve user experience. ---- I think it's a better way since the user is important to me.

Open to discussion.
